Addon Model & Bazarr Integration for Subtitles

Klubas opened the pull request:
**Basic bazarr integration for radarr.** 

- Now you can create addons for your services, in this case bazarr is an addon for radarr or sonarr 
- In the edit menu, now it's possible to search and download subtitles

**Config example**:

```
apis:
  movie:
    api_host: "http://radarr.localhost"
    api_key: "***"
  subtitles:
    api_host: "http://bazarr.localhost"
    api_key: "***"

services:
  - type: "Radarr"
    commands: ["movie", "m", "filme", "f"]
    api: "movie"
    name: "radarr"
    addons:
      - type: "Bazarr"
        service_name: "bazarr"
  - type: "Bazarr"
    commands: ["subtitles", "sub", "legenda", "l"]
    api: "subtitles"
    name: "bazarr"
```

**TODO:**
- [x] Allow user to go back to previous menu when searching for subtitles (I couldn't find a straight forward way to share the radarr state with the bazarr service)
- [ ] Show what subtitle is currently downloaded for each file (not sure if there is an api for this though)
- [x] Sonarr integration.
- [ ] Create bazarr service as standalone, in the current model it can only be called from a radarr/sonarr message.

**Bazarr standalone:**
- [ ] List wanted movies
- [ ] List wanted episodes
- [ ] Search wanted movie from wanted list
- [ ] Search wanted episode from wanted list
- [ ] Show providers status

TrimVis commented:
First of all, thanks for the PR!
Really looking forward to the result :)

> Allow user to go back to previous menu when searching for subtitles (I couldn't find a straight forward way to share the radarr state with the bazarr service)

In general it feels like inheritance could solve your issues :)
By inheriting from the Bazarr service, you can reuse its functionality and even call the corresponding super keyboard function etc directly from Radarr.
(This should be similar, to how I'm doing it with the ExtArrService)

Klubas commented:
Revert all changes, I'll be using another approach in the next commits, please start review here

TrimVis commented:
I really like the addon approach :)

Make sure to request a review from me, whenever you feel like this is close to being finished or you need some feedback :+1:

Klubas commented:
Hey @TrimVis , can you take a look at this?

I implemented the radarr and sonarr integration, not sure if I did it the best way though, your coding style left me a bit... confused sometimes hehe

Some commits may be kinda hard to follow, but basically I:
- Created an Addon class, servies that inherit from this class can be used as addons for other services
- Created and AddonState, it's and auxiliar state, inside the service state with the references to the service (and it's state) that called the addon service.
- Created an generic addonloader that needs to implemented in any service that supports addons `keyboard()` method
- Inside the addons there must be a `addon_buttons` method that returns the addon operations/buttons

Note 1: I implemented the seasons and episodes buttons inside sonarr, but I saw today that there's another PR from you with this functionality, these two probably could be refactored to use the same functions with some work when merging it, for now I'll leave it as is.

Note and help needed: 
To load addonState I had to use global variables inside the service (Addon.load() and Addon.config()), because I can't pass the object for  the callback functions directly (it's too largew iirc), and I _think_ this may cause some issues when there's too users using the bot at the same time

TrimVis commented on butlarr/config/services.py:
```diff
@@ -1,30 +1,54 @@
 SERVICES = []
 
 for service in CONFIG["services"]:
+    name = service.get("name")
```
Make sure to document this change in the README :)

TrimVis commented on butlarr/config/services.py:
```diff
@@ -1,30 +1,54 @@
-        assert False, "Unsupported service type!"
+        assert False, f"Unsupported service type {service_type}!"
+
+for service in SERVICES:
```
I don't like this loop and the access to SERVICES in the load_addons method. This is relatively convoluted and might be easy to break by accidentally changing a thing or two in another file.

Klubas replied:
I moved the logic to the services.py file, may need to optmize it a bit since I couldn't use your code suggestion as is but will take a look at it again later with more time to spare

TrimVis commented on butlarr/services/__init__.py:
```diff
@@ -2,11 +2,20 @@
 from loguru import logger
 from enum import Enum
 from typing import List, Tuple, Optional, Any
+from gzip import decompress
```
This import seems unused

Klubas replied:
done

TrimVis commented on butlarr/services/__init__.py:
```diff
@@ -66,7 +78,7 @@ def _delete(self, endpoint, params={}):
 
-    def request(self, endpoint: str, *, action=Action.GET, params={}, fallback=None):
+    def request(self, endpoint: str, *, action=Action.GET, params={}, fallback=None, raw=False):
         r = None
```
We should probably add an assert or something here to prevent users from using raw and fallback at the same time, as it would cause potentially unexpected behaviour :)

TrimVis commented on butlarr/services/ext.py:
```diff
@@ -112,3 +114,78 @@ async def cmd_help(self, update, context, args):
 
         return await update.message.reply_text(response_message, parse_mode="Markdown")
+    
+    def media_caption(self, item, overview=True):
```
The implementation of this should be moved to ArrService imo.
I would even argue that we should force users to implement it by throwing a NotImplemented Exception per default, but that's more a taste thing.
But this for sure shouldn't be part of the ExtService :)

Klubas replied:
Changed it too ArrService, but left the implementation there since the behavior should be the same for sonarr and radarr, if any service needs another behavior then it can be reimplemented there

TrimVis commented on butlarr/services/ext.py:
```diff
@@ -112,3 +114,78 @@ async def cmd_help(self, update, context, args):
+        return caption
+
+    def load_addons(self):
+        from ..config.services import SERVICES
```
I'm not a fan of inline imports, please move this to the top of the file :)

TrimVis commented on butlarr/services/ext.py:
```diff
@@ -112,3 +114,78 @@ async def cmd_help(self, update, context, args):
+        from ..config.services import SERVICES
+        logger.info(f"Loading {self.name} addons")
+        addons = []
+        for addon in self.addons:
```
I think we should not spread this behaviour in a fashion like this across files and do this (as you initially had it) all in config/services.py

TrimVis commented on butlarr/services/ext.py:
```diff
@@ -112,3 +114,78 @@ async def cmd_help(self, update, context, args):
+    menu: str = None
+
+    # Set the service and state that is loading this addon
+    def config(func):
```
How about addon_config and addon_load instead of having a class?

TrimVis commented on butlarr/services/ext.py:
```diff
@@ -112,3 +114,78 @@ async def cmd_help(self, update, context, args):
+    state: any = None
+    menu: str = None
+
+class Addon:
+    service: ArrService = None
+    state: any = None 
+    menu: str = None
```
It is confusing that this class does double duty by providing the static function wrap definitions and the new addon_buttons interface. Move the functions out of there.
Maybe I should think about splitting the ext folder up into different files for clarity too :)

Klubas replied:
The addon class is supposed to provide all the "extras" a service need to be also used as and addon (so, the buttons, the wrappers and the parent state) I don't really think that it really is a double duty.

The parent state thing for sure can be changed in some way, for now i replaced the 3 vars with a "parent" object:

```python
class ParentState:
    service: ArrService = None
    state: any = None
    menu: str = None

class Addon:
    parent: ParentState = None
    supported_services = []
```

TrimVis commented on butlarr/services/sonarr.py:
```diff
@@ -15,7 +15,7 @@
     default_session_state_key_fn,
 )
-from ..tg_handler.keyboard import Button, keyboard
+from ..tg_handler.keyboard import Button, keyboard as keyboard
```
```suggestion
from ..tg_handler.keyboard import Button, keyboard
```

TrimVis replied:
No need for that change :)

TrimVis commented on butlarr/services/sonarr.py:
```diff
@@ -379,6 +405,10 @@ async def clbk_update(self, update, context, args, state):
                     menu=None,
                 )
                 full_redraw = True
+            elif len(args) > 1 and not is_int(args[1]):
+                state = replace(state, menu=args[1])
+                if args[1] in ("seasons", "episodes", "episode"):
+                    allow_edit = False
```
What are you doing here? I dont' really get it tbh.
Why are you overloading goto like that?
Why not just a new command like goto_seasons or goto_episodes?

Klubas replied:
I just needed allow edit to be false when changing the keyboard layout

> Why not just a new command like goto_seasons or goto_episodes?

That would probably work to, i'll rewrite that later

Klubas replied:
Created new "goto_menu" command, and reverted the "goto" command to how it was :D

commit: https://github.com/TrimVis/butlarr/pull/15/commits/9fab45a40a51f98ea4205d398608197833d2dbe2

TrimVis commented on butlarr/services/sonarr.py:
```diff
@@ -447,3 +479,97 @@ async def clbk_cancel(self, update, context, args, state):
         self.remove(id=state.items[state.index].get("id"))
         return Response(caption="Series removed!")
+    
+    @repaint
```
Do you have screenshots available of these submenues and what the idea behind them is?

Is it just possible this way to select a season/episode and get direct control over that? By e.g. directly downloading it or downloading subtitles for just it?
I like that concept that seasons are also "navigatable" and I think this could be something really useful :)

Klubas replied:
Yes' that's the ideia, I should propably had included the screenshots in the request heh :D
In the first screen there's and view seasons and get subtitles button, currently they are the same just generated from different services (sonnar vs bazarr)

![image](https://github.com/user-attachments/assets/14b94bc4-7524-47f3-9cd2-3286596eb44b)
![image](https://github.com/user-attachments/assets/fafb08e4-3435-408b-847d-62144717791c)
![image](https://github.com/user-attachments/assets/8309b5ad-166f-4182-b873-5ca9309f5cc6)
![image](https://github.com/user-attachments/assets/7c412f1b-2570-4163-a906-1be95ca4e198)
![image](https://github.com/user-attachments/assets/77cbac76-838c-4198-bd9b-9add6c3bdfd3)
![image](https://github.com/user-attachments/assets/bab4bc58-39fe-4225-8c3c-0fa281dd2003)

And for movies:
![image](https://github.com/user-attachments/assets/93a21405-9305-4ba0-9b35-2068fde4c905)
![image](https://github.com/user-attachments/assets/945f3483-1f40-4405-9fd6-485910e78228)

TrimVis replied:
I really like that :)
I recently finished my own PR with a season selection, but we could probably easily replace the current behaviour with yours (its more intuitive anyways)

TrimVis commented on butlarr/services/bazarr.py:
```diff
@@ -0,0 +1,313 @@
+    ):
+        assert item, "Missing required arg! You need to provide a item!"
+
+        if service == ArrVariant.RADARR:
```
The service should provide functions for these behaviours.
You already got a Addon class, use it to force an interface for that :)

Klubas replied:
I'm not sure I follow here. 

The idea here is that the bazarr Service must provide all the functionality to integrate with bazarr and sonarr, because that's what bazarr, the application, do: You can't use bazarr as a standalone service. 

The addon class I created is used for services that can be used as addons (bazarr), not for services that have addons (radarr and sonarr). 

Of course we can change that, just not sure how you would do it though.

TrimVis replied:
No I dont' think we should change that, your approach is the correct way imo.
We don't want to start moving "addon functionality" into the sonarr and radarr implementations, it might just cause confusion down the line

TrimVis commented on butlarr/services/bazarr.py:
```diff
@@ -0,0 +1,313 @@
+        parent_state = state.parent.state
+        media_item = parent_state.items[parent_state.index]
+
+        if ArrVariant(parent_service.arr_variant) == ArrVariant.SONARR:
```
Same thing here, why are we doing this service specific logic here and not have a common function with an interface given by Addon

Klubas replied:
Same answer here :)

TrimVis commented on butlarr/services/bazarr.py:
```diff
@@ -0,0 +1,313 @@
+        in_library = "id" in item and item["id"]
+
+        buttons = []
+        if ArrVariant(self.service.arr_variant) == ArrVariant.RADARR:
+            downloaded = True if "movieFile" in item else False
+
+            if parent_state.menu == "add" and downloaded:
+                movieId = item.get("id")
+                
+                buttons.append(
+                    Button(
+                        f"Search for subtitles",
+                        self.get_clbk("list", movieId)
+                    ),
+                )
+
+        elif ArrVariant(self.service.arr_variant) == ArrVariant.SONARR:
+
+            if parent_state.menu == "add" and in_library:
+                buttons.append(
+                    Button(
+                        f"Get Subtitles",
+                        self.service.get_clbk("seasons")
+                    ),
+                )
+
+            elif parent_state.menu == "episode" and in_library:
+
+                episodeId = item['selectedEpisodeId']
+                episode = self.service.get_episode(episodeId)
+                downloaded = True if episode['hasFile'] else False
+
+                if downloaded:
+                    buttons.append(
+                        Button(
+                            f"Search for subtitles",
+                            self.get_clbk("list", episodeId)
+                        ),
+                    )
```
Same here, use the Addon interface to force an implementation for this :)

Klubas replied:
I moved it to service especific functions, so it won't be as ugly but it's still inside bazarr.py for now

TrimVis reviewed (changes requested):
It took me some time, but I finally got to reviewing your PR :)
First of all, thanks for that! Really looking forward to the bazarr integration :)

It's a lot of comments, but most of it is just minor restructuring of your code (I.e. moving code to new functions in your addons class and stuff like that).

TrimVis commented:
@Klubas I just saw your note regarding the addon load logic, I will think about a way to optimize this and send it to you in some way :)

TrimVis commented:
> Note and help needed: To load addonState I had to use global variables inside the service (Addon.load() and Addon.config()), because I can't pass the object for the callback functions directly (it's too largew iirc), and I _think_ this may cause some issues when there's too users using the bot at the same time

I misread your last note and thought you had issues with relying on SERVICES with addons :D
Apparently you have fixed that issue, although I don't believe it would have caused any real issues with multiple users :)


Still, I took to the time to slightly refactor the `buttlarr/config/services.py` file, so that the load function is no longer required:

```
import importlib

from . import CONFIG

APIS = CONFIG["apis"]
SERVICES = []

# Keep the namespace clean
def _load_services():
    # Two step approach, due to addons requiering all services to be loaded beforehand
    named_services = {}
    for service in CONFIG["services"]:
        try:
            service_module = importlib.import_module(f"butlarr.services.{service["type"].lower()}")
            ServiceConstructor = getattr(service_module, service["type"])
        except Exception:
            assert False, "Could not find a module for that service"

        api_config = APIS[service["api"]]
        service_name = service.get("name")
        args = {
            "name": service_name,
            "commands": service["commands"],
            "api_host": api_config["api_host"],
            "api_key": api_config["api_key"],
            "addons": service.get("addons", []),
        }

        service = ServiceConstructor(**args)
        SERVICES.append(service)
        if service_name:
            assert service_name not in named_services, "Different services have been named the same!"
            named_services[service_name] = service

    for service in SERVICES:
        if not len(service.addons): continue

        logger.info(f"Loading addons for {self.name}")
        addons = []
        for required_service in self.addons:
            addon_service = named_services.get(required_service)
            addons.append(addon_service)
            logger.info(f"Addon {addon_service.name} loaded")
        service.addons = addons
        logger.debug(f"{self.name} service loaded Addons: {str(self.addons)}")

_load_services()

```

(NOTE that this code is mostly untested, but should work fine :))

Klubas commented:
> Still, I took to the time to slightly refactor the buttlarr/config/services.py file, so that the load function is no longer required:

I like this and saw that you already updated main to use this method :D

> and I think this may cause some issues when there's too users using the bot at the same time

About this, I ran some tests and the behaviar can be a bit unpredictable because of the global variables. I'm thinking about instantiating the Addon only when the parent service call for it instead of having only one instance running all the time.

TrimVis commented:
From a few glances, the code looks good now :)

I will thoroughly test this over the week and finish the final review; in case I don't encounter any issues, this should be pretty finalized :)

(Maybe some adaptions to allow backward compatibility with the config file format are necessary and probably a little merging, but nothing too major)

TrimVis commented:
> About this, I ran some tests and the behaviar can be a bit unpredictable because of the global variables. I'm thinking about instantiating the Addon only when the parent service call for it instead of having only one instance running all the time.

How did you test this? Simply trying to interact with the bot with multiple users at the same time?
A few more details for me to watch out for during testing would be good :)
I don't yet see where the addon instatiation could lead to errors with multiple users, so I want to be extra careful about it :D

Klubas commented:
I actually just sent two messages from the same chat, with different parameters. 

This may be the cause of the problem

Em 26 de agosto de 2024 17:48:43 BRT, TrimVis ***@***.***> escreveu:
>> About this, I ran some tests and the behaviar can be a bit unpredictable because of the global variables. I'm thinking about instantiating the Addon only when the parent service call for it instead of having only one instance running all the time.
>
>How did you test this? Simply trying to interact with the bot with multiple users at the same time?
>A few more details for me to watch out for during testing would be good :)
>I don't yet see where the addon instatiation could lead to errors with multiple users, so I want to be extra careful about it :D
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/TrimVis/butlarr/pull/15#issuecomment-2311065925
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>

TrimVis commented:
Moved to #23
